User management

TheJarles opened the pull request:
With this pull request comes the implementation of the first few tasks in milestone 3 in the project backlog. I've added basic username validation as well as password generation. The encryption is still WIP, but it's implemented using openssl methods that should also be available in PHP. A thing to note is that I've for simplicity implemented most of these functions in the User model and not in its respective controller as suggested in the projects cards.
I also finished up the database seeds which should now be finished.

jabbors commented on app/models/user.rb:
```diff
@@ -1,2 +1,36 @@
+  validates :username, uniqueness: true
+  def generate_password
+    clear_pass = password.nil? || password.empty? ? Sysrandom.base64(9) : password
+    key = OpenSSL::PKey::RSA.new EXAMPLE_RSA_KEY
```
It works, but no. My task definition was maybe not very clear and I'm sorry for that. I'm looking for a simpler encryption solution that uses a predefined `secret` for encryption and decryption. 

Something more like this https://gist.github.com/0sc/2e2bdb19bcc0051e715625075e7cc068. But even simpler. In the gist the key is generated from random salts whereas we need to use a fixed key which is passed to the Rails environment through e.g. the `SECRET_KEY` environment variable.

This secret shall in turn be used in KADOS to decrypt the base64 encoded and encrypted password stored in the database.

The secret has to be exactly 32 char long for AES encryption but I've usually cheated and generated a sha256 from the secret to obtain a 32 char long secret. Hence the user can decide to use a short and bas password like `pass123` or a long random password like `NmQxYmIwN2Y1NjgxNGRmYWYzN2FkOTY4YTJiY2VkMjRmMDFlNmNjNGZlY2M2ZTE4` and I always get the fixed 32 char secret.

I recently picked AES-256-GCM as a cipher in another project I created and if you have time you could explore if this alternative works in both Ruby and PHP.

jabbors replied:
I experimented with AES-256-GCM yesterday and found that Ruby, PHP and Golang all seems to implement encryption/decryption differently. So far I've only been able to encrypt and decrypt a secret within one language. Trying to encrypt in e.g. Ruby and decrypt in PHP is still a mystery how to get it working.

To be continued ...

jabbors replied:
So now I managed to encrypt with Ruby and decrypt with PHP.

```ruby

jabbors replied:
The catch was to find the right option in PHP. I figured OPENSSL_RAW_DATA was the default so  tested a bunch of other flags and combination. Then I decided to try OPENSSL_RAW_DATA and got it working 🎉

TheJarles replied:
All right, I'll switch the encryption to AES-GCM. I'll have to look around a bit how to implement secrets in rails and I'll do that aswell.

jabbors replied:
With docker and kubernetes secrets and other configuration parameters are given as environment variables, thus you just have to use `ENV["SOME_VARIABLE_NAME"]` in your code.

TheJarles commented on db/seeds.rb:
```diff
@@ -6,8 +6,19 @@
+                       { state: 'RESUME' }, { state: 'FINISH' }])
+content_lists = ContentList.create([{ name: 'bookshelf' }, { name: 'new' },
+                                    { name: 'issued' }, { name: 'expired' },
+                                    { name: 'search' }, { name: 'browse' }])
```
I'll look into it. It probably is possible to do.

TheJarles commented:
I implemented the AES-256-GCM encryption as requested. I made some minor tweaks to your example code. I removed the assignment of cipher.iv to iv as it's redundant, the call to random_iv already does the assignment internally. Other than that I also added an assignment of auth_data to '' as reading through the ruby OpenSSL docs revealed this is necessary. I tested locally and the encryption seems to work as intended, I did not test Ruby to PHP encryption/decryption though, but seeing as the code is quite similar to yours I'd wager it works. Finally I set up the SECRET_KEY environment variable and made the code use that for the encryption.
  Regarding the manual id assignments to the ContentList entries, the solution is not necessarily the most efficient (it requires separate db lookup) but it should work without any hickups. I stumbled upon a couple of alternatives, but it seems this way was the simplest and least bug-prone way to do it.

jabbors commented:
I ensured that we can decrypt the passwords in PHP after your modifications. 👍 

And the custom id assignment is good enough. All good for merge.
